module_utils/network_lsr/ethtool.py

```diff
@@ -0,0 +1,54 @@
+# SPDX-License-Identifier: BSD-3-Clause
+
+import array
+import binascii
+import struct
+import fcntl
+import socket
+
+
+ETHTOOL_GPERMADDR = 0x00000020
+SIOCETHTOOL = 0x8946
+
+
+def get_perm_addr(ifname):
+    """
+    Return the Permanent address value for the specified interface using the
+    ETHTOOL_GPERMADDR ioctl command.
+
+    Please for further documentation, see:
+    https://github.com/torvalds/linux/blob/master/include/uapi/linux/ethtool.h#L734
+    https://github.com/torvalds/linux/blob/master/include/uapi/linux/ethtool.h#L1388
+    https://git.kernel.org/pub/scm/network/ethtool/ethtool.git/tree/ethtool.c#n4172
+    """
+    try:
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sockfd = sock.fileno()
+        ifname = ifname.encode("utf-8")
+        ecmd = array.array("B", struct.pack("II6s", ETHTOOL_GPERMADDR, 6, b"\x00" * 6))
+        ifreq = struct.pack("16sP", ifname, ecmd.buffer_info()[0])
+
+        fcntl.ioctl(sockfd, SIOCETHTOOL, ifreq)
+        try:
+            res = ecmd.tobytes()
+        except AttributeError:
+            # tobytes() is not available in python2
+            res = ecmd.tostring()
+        _, _, perm_addr = struct.unpack("II6s", res)
+        perm_addr = _decode_perm_addr(perm_addr)
+    except IOError:
+        perm_addr = None
+    finally:
+        sock.close()
+
+    return perm_addr
+
+
+def _decode_perm_addr(perm_addr_bytes):
+    if hasattr(perm_addr_bytes, "hex"):
+        perm_addr = perm_addr_bytes.hex()
+    else:
+        perm_addr = binascii.hexlify(perm_addr_bytes)
```

Comment:
hexlify seems to be available in all Python versions, can't you use it all the time?

Reply:
No I cannot use it all the time. `binascii.hexlify()` is returning a bytes representation instead of str in python3 [1]. In order to make this work I would need to parse the bytes representation into a string. This is what hex() does.

Using it all the time would generate another conflict.. so I don't think we should use it all the time.

[1] https://docs.python.org/3.6/library/binascii.html

Reply:
I see, can you please use try/except here, too, and mention this in a comment that mentions Python 2. Thanks.

Reply:
After looking at `Util.mac_ntoa()`, can't you use something like `":".join(["%02x" % byte for byte in perm_addr_bytes])` or actually decode the MAC using `Util.mac_ntoa()` in the module?

Reply:
I cannot do that. `":".join(["%02x" % byte for byte in perm_addr_bytes])` only works for bytes representation and there is no such thing in python2. Here is the error that you get on python2 using it:

```
    @staticmethod
    def mac_ntoa(mac):
        if mac is None:
            return None
>       return ":".join(["%02x" % c for c in mac])
E       TypeError: %x format: a number is required, not str
```

Reply:
The `mac_ntoa()` requires a byte representation as input. This function is being used in the validation because the address is being converted to byte representation before using `mac_aton()` [1]. In this case, we do have different types in python2 (str) and python3 (bytes).

[1] https://github.com/linux-system-roles/network/blob/master/module_utils/network_lsr/argument_validator.py#L302

Reply:
Ok, thank you. So the problem is that `struct.unpack()` returns `str` in Python2 and `bytes` in Python3. It seems that bytearray helps here:

```
$ python3 -c 'print(":".join(["%02x" % byte for byte in bytearray(b"abcdef")]))'
61:62:63:64:65:66
$ python2 -c 'print(":".join(["%02x" % byte for byte in bytearray("abcdef")]))'
61:62:63:64:65:66
```

What do you think?

Reply:
Oh thanks! That works :-). I've update the commit